packages/org.openapis.v3/OAuthFlows.pkl

```diff
@@ -0,0 +1,35 @@
+//===----------------------------------------------------------------------===//
+// Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+/// Specific OAuth flow types.
+module org.openapis.v3.OAuthFlows
+
+import "OAuthFlow.pkl"
+
+/// Configuration for the OAuth Implicit flow.
+implicit: OAuthFlow.WithAuthorizationUrl
+
+/// Configuration for the OAuth Resource Owner Password flow.
+password: OAuthFlow.WithTokenUrl
+
+/// Configuration for the OAuth Client Credentials flow.
+///
+/// Previously called application in OpenAPI 2.0.
+clientCredentials: OAuthFlow.WithTokenUrl
+
+/// Configuration for the OAuth Authorization Code flow.
+///
+/// Previously called accessCode in OpenAPI 2.0.
+authorizationCode: OAuthFlow.WithTokenAndAuthorizationUri
```

Comment:
It's a little nicer here to just have a single class with some constraints. This means that in the usage site, you don't need to figure out how to import the correct type.

```suggestion

/// Configuration for the OAuth Implicit flow.
implicit: OAuthFlow(authorizationUrl != null)?

/// Configuration for the OAuth Resource Owner Password flow.
password: OAuthFlow(tokenUrl != null)?

/// Configuration for the OAuth Client Credentials flow.
///
/// Previously called application in OpenAPI 2.0.
clientCredentials: OAuthFlow(tokenUrl != null)?

/// Configuration for the OAuth Authorization Code flow.
///
/// Previously called accessCode in OpenAPI 2.0.
authorizationCode: OAuthFlow(tokenUrl != null && authorizationUrl != null)?

class OAuthFlow {
  /// The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL.
  refreshUrl: Uri?

  /// The available scopes for the OAuth2 security scheme.
  ///
  /// A map between the scope name and a short description for it. The map MAY be empty.
  scopes: Mapping<String, String>

  /// The authorization URL to be used for this flow.
  ///
  /// This MUST be in the form of a URL.
  authorizationUrl: Uri?

  /// The token URL to be used for this flow.
  ///
  /// This MUST be in the form of a URL.
  tokenUrl: Uri?
}
```